JavaScript/JavaScriptSDK.Tests/CheckinTests/Sender.tests.ts

```diff
@@ -655,12 +666,497 @@ class SenderTests extends TestClass {
+
+                var delay = 27; // in seconds
+                var delayDate = new Date();
+                delayDate.setSeconds(delayDate.getSeconds() + delay);
+
+                (<any>sender)._setRetryTime(delayDate.toUTCString());
+
+                Assert.equal(now + delay * 1000, (<any>sender)._retryAt, "Invalid retry time.");
+            }
+        });
     }
 
     private setupDataLossAnaluzed() {
         Microsoft.ApplicationInsights.DataLossAnalyzer.enabled = true;
         Microsoft.ApplicationInsights.DataLossAnalyzer.appInsights = <any>{ trackTrace: (message) => { }, flush: () => { }, context: { _sender: { _XMLHttpRequestSupported: true } } };
     }
+
+    private validatePartialSuccess_NonRetryable() {
+        var sender = this.getSender();
+        Assert.ok(sender, "sender was constructed");
+
+        // send two items
+        this.fakeServer.requests.pop();
+        sender.send(this.testTelemetry);
+        sender.send(this.testTelemetry);
+
+        Assert.equal(2, sender._buffer.count(), "Buffer has two items");
+
+        // trigger send
+        this.clock.tick(sender._config.maxBatchInterval());
+
+        this.requestAsserts();
+        this.fakeServer.requests.pop().respond(
+            206,
+            { "Content-Type": "application/json" },
+            // backend rejected 1 out of 2 payloads. First payload was too old and should be dropped (non-retryable).
+            '{ "itemsReceived": 2, "itemsAccepted": 1, "errors": [{ "index": 0, "statusCode": 400, "message": "103: Field time on type Envelope is older than the allowed min date. Expected: now - 172800000ms, Actual: now - 31622528281ms" }] }'
+        );
+
+        // verify
+        Assert.ok(sender.successSpy.called, "success was invoked");
+
+        this.logAsserts(1);
+        Assert.equal('AI (Internal): NONUSRACT_OnError message:"Failed to send telemetry." props:"{message:partial success 1 of 2}"', this.loggingSpy.args[0][0], "Expecting one warning message");
+
+        // the buffer is empty. 
+        Assert.equal(0, sender._buffer.count(), "Buffer is empty");
+
+        // session storage buffers are also empty
+        var buffer: string[] = JSON.parse(Microsoft.ApplicationInsights.Util.getSessionStorage(Microsoft.ApplicationInsights.SessionStorageSendBuffer.BUFFER_KEY));
+        var sentBuffer: string[] = JSON.parse(Microsoft.ApplicationInsights.Util.getSessionStorage(Microsoft.ApplicationInsights.SessionStorageSendBuffer.SENT_BUFFER_KEY));
+
+        Assert.equal(0, buffer.length, "Session storage buffer is empty");
+        Assert.equal(0, sentBuffer.length, "Session storage sent buffer is empty");
+
+        // clean up
+        sender.successSpy.reset();
+        sender.errorSpy.reset();
+    }
+
+    private validatePartialSuccess_Retryable() {
+        var sender = this.getSender();
+        Assert.ok(sender, "sender was constructed");
+
+        // send six items
+        this.fakeServer.requests.pop();
```

Comment:
Sinon library puts an empty request on the stack. I was trying to figure out why, but I gave up.